File: programmers/js/level01/160586.js
// 대충 만든 자판
function solution(keymap, targets) {
  const answer = [];
  const countKeyObj = countKeymap(keymap);

  for (const target of targets) {
    let count = 0;
    
    for (const char of target) {
      if (!(char in countKeyObj)) {
        answer.push(-1);
        break;
      }
      
      count += countKeyObj[char];
    }
    
    if (count !== 0) {
      answer.push(count);
    }
  }
  
  return answer;
}

function countKeymap(keymap) {
  const keymapSet = new Set([...keymap.join('')]);
  const keymapObj = {};
  
  keymapSet.forEach(k => {
    keymap.forEach(key => {
      if (key.includes(k)) {
        const kIndex = key.indexOf(k);
        keymapObj[k] = keymapObj[k] > kIndex ? 
          kIndex + 1 : (keymapObj[k] || kIndex + 1);
      }
    });
  });
  
  return keymapObj;
}

File: programmers/java/level01/160586.java
// 대충 만든 자판

import java.util.ArrayList;

class Solution {
    int getKeymapCount(String[] keymap, char c) {
        int count = 101;
        
        for (String key : keymap) {
            int idx = key.indexOf(c);
            
            if (idx == -1) continue;
            
            count = Math.min(count, idx + 1);
        }
        
        return count == 101 ? -1 : count;
    }
    
    public int[] solution(String[] keymap, String[] targets) {
        ArrayList<Integer> result = new ArrayList<>();
        
        for (String target : targets) {
            int count = 0;
            
            for (char c : target.toCharArray()) {
                int temp = getKeymapCount(keymap, c);
                
                if (temp == -1) {
                    count = -1;
                    break;
                }
                
                count += temp;
            }
            
            result.add(count);
        }
        
        return result.stream().mapToInt(v -> v).toArray();
    }
}